airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py

```diff
@@ -0,0 +1,337 @@
+                slice_range=step_length,
+                lookback_window=parse_duration(declarative_cursor.lookback_window) if declarative_cursor.lookback_window else None,
+                cursor_granularity=parse_duration(declarative_cursor.cursor_granularity) if declarative_cursor.cursor_granularity else None,
+            )
+        return None
+
+    @staticmethod
+    def _stream_supports_concurrent_partition_processing(declarative_stream: DeclarativeStream, cursor: DatetimeBasedCursor) -> bool:
+        """
+        Many connectors make use of stream_state during interpolation on a per-partition basis under the assumption that
+        state is updated sequentially. Because the concurrent CDK engine processes different partitions in parallel,
+        stream_state is no longer a thread-safe interpolation context. It would be a race condition because a cursor's
+        stream_state can be updated in any order depending on which stream partition's finish first.
+
+        We should start to move away from depending on the value of stream_state for low-code components that operate
+        per-partition, but we need to gate this otherwise some connectors will be blocked from publishing. See the
+        cdk-migrations.md for the full list of connectors.
+        """
+
+        # If a stream doesn't partition the sync time window into smaller intervals, then we can sync this concurrently because
+        # there is not a race condition on a single partition.
+        if not cursor.step:
+            return True
+
+        if isinstance(declarative_stream.retriever, SimpleRetriever) and isinstance(declarative_stream.retriever.requester, HttpRequester):
+            http_requester = declarative_stream.retriever.requester
+            if "stream_state" in http_requester._path.string:
+                return False
+
+            request_options_provider = http_requester._request_options_provider
+            if request_options_provider.request_options_contain_stream_state():
+                return False
+
+            record_selector = declarative_stream.retriever.record_selector
+            if isinstance(record_selector, RecordSelector):
+                if record_selector.record_filter and "stream_state" in record_selector.record_filter.condition:
+                    return False
+
+                for add_fields in [
+                    transformation for transformation in record_selector.transformations if isinstance(transformation, AddFields)
+                ]:
+                    for field in add_fields.fields:
+                        if isinstance(field.value, str) and "stream_state" in field.value:
+                            return False
+                        if isinstance(field.value, InterpolatedString) and "stream_state" in field.value.string:
+                            return False
+        return True
+
+    @staticmethod
+    def _get_cursor(stream: DeclarativeStream) -> Optional[DeclarativeCursor]:
```

Comment:
The concern I have with this method is that we are coupling the creation of the concurrent cursor with code that should eventually be removed i.e. we do (declarative -> low-code component -> concurrent) instead of (declarative -> concurrent). Once we will remove the low-code components, we will have to re-implement this logic. Should we instead rely on `airbyte_cdk.sources.declarative.models.declarative_component_schema.DatetimeBasedCursor`?

Reply:
That is a fair question. I think this can get a little bit awkward because we need to sort of extract the DatetimeBasedCursor model from the `_source_config` field and we lose the automatic parsing into interpolated fields and other things that we basically need to reimplement. But it does feel like the model is the more solid interface that we won't change. I will play around with this tomorrow and see if this can be done sanely